src/Helpers.dfy

```diff
@@ -0,0 +1,300 @@
+    (forall x, y :: less(x, y) && less(y, x) ==> false) &&  // not both of the less's
+    (forall x, y :: less(x, y) ==> x != y)  // not a less and the equality
+  }
+
+  predicate Transitive<T(!new)>(R: (T, T) -> bool) {
+    forall x, y, z :: R(x, y) && R(y, z) ==> R(x, z)
+  }
+
+  /*
+   * Here is an example relation and a lemma that says the relation is appropriate for use in
+   * lexicographic orderings.
+   */
+
+  lemma UInt8LessIsTrichotomousTransitive()
+    ensures Trichotomous(UInt8Less)
+    ensures Transitive(UInt8Less)
+  {
+  }
+
+  /*
+   * As the following lemmas show, the lexicographic ordering is reflexive, antisymmetric, transitive, and total.
+   * The proofs are a bit pedantic and include steps that can be automated.
+   */
+
+  lemma LexIsReflexive<T>(a: seq<T>, less: (T, T) -> bool)
+    ensures LexicographicLessOrEqual(a, a, less)
+  {
+    assert LexicographicLessOrEqualAux(a, a, less, |a|);
+  }
+
+  lemma LexIsAntisymmetric<T>(a: seq<T>, b: seq<T>, less: (T, T) -> bool)
+    requires Trich: Trichotomous(less)
+    requires LexicographicLessOrEqual(a, b, less)
+    requires LexicographicLessOrEqual(b, a, less)
+    ensures a == b
+  {
+    assert LessIrreflexive: forall x,y :: less(x, y) ==> x != y by {
+      reveal Trich;
+    }
+    assert ASymmetric: forall x,y :: less(x, y) && less(y, x) ==> false by {
+      reveal Trich;
+    }
+    var k0 :| 0 <= k0 <= |a| && LexicographicLessOrEqualAux(a, b, less, k0);
+    var k1 :| 0 <= k1 <= |b| && LexicographicLessOrEqualAux(b, a, less, k1);
+    var max := if k0 < k1 then k1 else k0;
+    assert max <= |a| && max <= |b|;
+    assert SameUntilMax: forall i :: 0 <= i < max ==> a[i] == b[i];
+    assert AA: k0 == |a| || (k0 < |b| && less(a[k0], b[k0]));
+    assert BB: k1 == |b| || (k1 < |a| && less(b[k1], a[k1]));
+    calc {
+      true;
+    ==>  { reveal AA, BB; }
+      (k0 == |a| || (k0 < |b| && less(a[k0], b[k0]))) && (k1 == |b| || (k1 < |a| && less(b[k1], a[k1])));
+    ==  // distribute && and ||
+      (k0 == |a| && k1 == |b|) ||
+      (k0 == |a| && k1 < |a| && less(b[k1], a[k1])) ||
+      (k0 < |b| && less(a[k0], b[k0]) && k1 == |b|) ||
+      (k0 < |b| && less(a[k0], b[k0]) && k1 < |a| && less(b[k1], a[k1]));
+    ==  { reveal LessIrreflexive, SameUntilMax; }
+      (k0 == |a| && k1 == |b|) ||
+      (k0 < |b| && less(a[k0], b[k0]) && k1 < |a| && less(b[k1], a[k1]));
+    ==>  { reveal LessIrreflexive, SameUntilMax; assert max <= k0 && max <= k1; }
+      (k0 == |a| && k1 == |b|) ||
+      (k0 < |b| && less(a[k0], b[k0]) && k1 < |a| && less(b[k1], a[k1]) && k0 == k1 == max);
+    ==  { reveal ASymmetric; }
+      k0 == |a| && k1 == |b|;
+    ==>  { assert |a| == k0 <= max && |b| == k1 <= max ==> k0 == k1; }
+      max == |a| == |b|;
+    ==>  { reveal SameUntilMax; }
+      a == b;
+    }
+  }
+
+  lemma LexIsTransitive<T>(a: seq<T>, b: seq<T>, c: seq<T>, less: (T, T) -> bool)
```

Comment:
Can you move this and everything Lexicographic related to the lexicographics module?
You could always define another module, like Lexicographics.ProofHelpers

Also, please rename "less" to "lessThan".
Consider (but think carefully if it will be better) to change "less" to "compare" that returns a ComparisonResult, for consistentcy